### PR TITLE
added /ready and /ready/sync endpoints

### PR DESCRIPTION
**Pull request by onuruci**
Implements #2052

Added /ready and /ready/sync endpoints to jsonrpcServer.

/ready endpoint return 200 http code if jsonrpcServer is ready to be used and returns 503 code else.

/ready/sync first does the same check with /ready endpoint and return 503 if not ready. If node is synced within last 6 blocks returns 200 http code and if not returns 503.

### CODE REVIEW

File: rpc/monitor.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	}
+	highestBlockHeader := h.syncReader.HighestBlockHeader()
+	if highestBlockHeader == nil {
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	}
+
+	if head.Number+6 >= highestBlockHeader.Number {
```
Comment (kirugan) [inline rpc/monitor.go:31]: let's move magic number (`6`) to a package level constant in this file

Comment (onuruci) [reply rpc/monitor.go]: moved in the new commit

File: rpc/monitor.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+func (h *Handler) HandleReadyRequest(w http.ResponseWriter, r *http.Request) {
+	if h.ready {
+		w.WriteHeader(http.StatusOK)
+	} else {
+		w.WriteHeader(http.StatusServiceUnavailable)
+	}
+}
+
```
Comment (rianhughes) [inline rpc/monitor.go:18]: nit:
func (h *Handler) HandleReadyRequest(w http.ResponseWriter, r *http.Request) {
	if h.ready {
		w.WriteHeader(http.StatusOK)
	}
	w.WriteHeader(http.StatusServiceUnavailable)
}

File: node/node.go [hunk 3]
```diff
@@ -271,6 +272,8 @@ func New(cfg *Config, version string) (*Node, error) { //nolint:gocyclo,funlen
 		log.Warnw("Failed to parse Juno version, will not warn about new releases", "version", version)
 	}
 
+	rpcHandler.SetReady()
```
Comment (rianhughes) [inline node/node.go:275]: I wonder if we should set this in rpcHandler.Run()??

Comment (onuruci) [reply node/node.go:275]: Yes I think that will be better

Comment (onuruci) [reply node/node.go]: Updated this to setting rpcHandler ready in rpcHandler.run and updated tests so that they check readiness after rpcHandler.run

**Review by rianhughes — APPROVED** [review]

File: node/http.go [hunk 4]
```diff
@@ -96,7 +102,7 @@ func makeRPCOverHTTP(host string, port uint16, servers map[string]*jsonrpc.Serve
 }
 
 func makeRPCOverWebsocket(host string, port uint16, servers map[string]*jsonrpc.Server,
-	log utils.SimpleLogger, metricsEnabled bool, corsEnabled bool,
+	log utils.SimpleLogger, metricsEnabled bool, corsEnabled bool, rpchandler *rpc.Handler,
```
Comment (kirugan) [inline node/http.go:105]: the reason why we need `/ready` and `/ready/sync` paths are readiness checks in kubernetes kind of environment. There is no need for now to have it on websocket. So I propose to revert this changes for websockets

File: node/http.go [hunk 5]
```diff
@@ -86,6 +87,11 @@ func makeRPCOverHTTP(host string, port uint16, servers map[string]*jsonrpc.Serve
 			httpHandler = httpHandler.WithListener(listener)
 		}
 		mux.Handle(path, exactPathServer(path, httpHandler))
+
+		prefixedPath := strings.TrimSuffix(path, "/")
+
+		mux.HandleFunc(prefixedPath+"/ready", rpchandler.HandleReadyRequest)
+		mux.HandleFunc(prefixedPath+"/ready/sync", rpchandler.HandleReadySyncRequest)
```
Comment (kirugan) [inline node/http.go:94]: I think the best way to structure it would be: register `/ready` path in special http readiness handler (see `servers` variable). It would give us these benefits:

1. reduce number of readiness paths
2. reuse same logic for path handling assignments
3. separation of conerns: jsonrpc server struct should handle only corresponding requests and separate structure for readiness checks should handle its own kind of requests

wdyt?

File: node/http.go (hunk 4 above)
Comment (onuruci) [reply node/http.go]: Removed it in 294c5eefcfc3ecd6465dbde1be62b0d5f8497314

File: Makefile [hunk 6]
```diff
@@ -15,7 +15,6 @@ ifeq ($(shell uname -s),Darwin)
 
 	# Set macOS deployment target in order to avoid linker warnings linke
 	# "ld: warning: object file XX was built for newer macOS version (14.4) than being linked (14.0)"
-	export MACOSX_DEPLOYMENT_TARGET=$(shell sw_vers --productVersion)
```
Comment (kirugan) [inline Makefile:18]: removed by mistake?

Comment (onuruci) [reply Makefile:18]: Yes

File: node/http.go (hunk 5 above)
Comment (onuruci) [reply node/http.go]: This sounds like the correct way. I separated the logic from jsonrpc server and moved it into a special http handler. Reduced the number of paths serving for readiness.

File: jsonrpc/readiness.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+func NewReadinessHandlers(bcReader blockchain.Reader, syncReader sync.Reader) *ReadinessHandlers {
+	return &ReadinessHandlers{
+		bcReader:   bcReader,
+		syncReader: syncReader,
+	}
+}
+
+func (h *ReadinessHandlers) HandleReady(w http.ResponseWriter, _ *http.Request) {
+	w.WriteHeader(http.StatusOK)
+}
```
Comment (IronGauntlets) [inline jsonrpc/readiness.go:26]: Since this function always returns http.StatusOK, do you think it is a good idea to inline this function on line 215 in node/node.go?

File: jsonrpc/readiness.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+	if head.Number > highestBlockHeader.Number {
+		return false
+	}
+
+	if head.Number+SyncBlockRange >= highestBlockHeader.Number {
+		return true
+	} else {
+		return false
+	}
```
Comment (IronGauntlets) [inline jsonrpc/readiness.go:55]:
```suggestion
	if head.Number+SyncBlockRange >= highestBlockHeader.Number {
		return true
	}
	return false
```

Comment (kirugan) [reply jsonrpc/readiness.go:55]: good point, do you think it will be more readable as this:
```
return head.Number+SyncBlockRange >= highestBlockHeader.Number
```

File: jsonrpc/readiness.go
Comment (IronGauntlets) [inline jsonrpc/readiness.go:1]: I think these files should not be under the node package, I don't think it is a good idea for jsonrpc package to import blockchain and sync packages.

File: jsonrpc/readiness.go (hunk 8 above)
Comment (IronGauntlets) [reply jsonrpc/readiness.go:55]: Yes, that is even better.

File: jsonrpc/readiness.go
Comment (IronGauntlets) [reply jsonrpc/readiness.go:1]: Potentially under `node/http.go`file.

File: jsonrpc/readiness.go (hunk 7 above)
Comment (kirugan) [reply jsonrpc/readiness.go:26]: I asked about `/ready` endpoint in [tg](https://t.me/StarknetJuno/1224). Seems like we can safely remove it and just rely on / root path.

File: jsonrpc/readiness.go
Comment (kirugan) [reply jsonrpc/readiness.go:1]: >  I don't think it is a good idea for jsonrpc package to import blockchain and sync packages.

good point, I'll move them to `http.go` are you ok with `ReadinessHandlers` structure though? + most likely I'll make it unexported

**Review by IronGauntlets — APPROVED** [review]
LGTM
